### PR TITLE
Implementation Plan: Documenting Media properties in the catalog

### LINKED ISSUE #1995: Implementation Plan: Clearly document all media properties in the catalog

### PR DESCRIPTION
**Pull request by obulat**
## Due date:

2023-05-15

## Assigned reviewers

- [x] @AetherUnbound 
- [x] @stacimc 

I've added you because of your extensive work on the catalog.
Also tagging @rwidom for her deep insights.

## Description

Fixes #1995 by @obulat 
Implementation plan for project #412.

## Current round

This discussion is following the Openverse decision-making process. Information about this process can be found [on the Openverse documentation site](https://docs.openverse.org/reference/decision_making/index.html). Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

This discussion is currently in the Decision round.

The deadline for review for this round is May 15, 2023 (Monday).

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
+
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :59]: Does any of the validation in the `MediaStore` need to be captured, or is that all captured in the `Column` definitions? (For example I do think that all the fields checked as required fields in the MediaStore are actually non-nullable)

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+### 3. Adding the long-form information from the markdown file
+The markdown file will be parsed and split into the preamble, postamble (???) and the information about each data property. This information would be added to the `description` in the property mapping.
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :61]: Which markdown file is this referring to? Also, TIL `postamble` 😄

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+### Information to include in `media_properties.md` for each property
+Media types: Image, Audio
+Short description
+Names used in the provider scripts, if different from the database name
+Shape of the object (if it is not a simple type)
+Selection criteria
+Normalization and validation performed in the `MediaStore` class and the relevant `Column`
+Inconsistencies in the database data
+See sample for `identifier` and `tags` in `sample_media_properties.md`
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :74]: You mentioned earlier that some of this data would need to be updated manually. Do we know which parts?

I have a couple of questions about the manual input as well:
* How will we ensure that the `just` script doesn't overwrite that data when it auto-generates the documentation?
* How will we ensure that this gets updated as fields change or are added? Will this just be a checklist item on the PR list, for example, or did you have some automation in mind?

File: documentation/projects/proposals/document_all_media_properties/sample_media_properties.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Media Properties
+
+_This document is auto-generated from the source code in
+`utilities/media_props_gen/media_props_generation.py`. _
+
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/sample_media_properties.md:5]: I really love that you included this sample documentation! 🤩

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
+### 3. Adding the long-form information from the markdown file
+The markdown file will be parsed and split into the preamble, postamble (???) and the information about each data property. This information would be added to the `description` in the property mapping.
+
+### 4. Writing the markdown file
+The mapping will be used to create a new markdown file: the Preamble, the tables for Audio and Image, the long-form document from step 3.
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :64]: I think I'm confused about the files written here. In the overview you wrote:


> The information on data types, transformations, and validations will be presented as a table and will be extracted from the code.  Other information about the data selection properties and data discrepancies will be written in a separate markdown file as it is easier to read, write, and lint.

Is step 3 referring to generating a long form file with 'other information' (the second thing listed above), and step 4 is referring to a separate document with just the tables?

> The mapping will be used to create a new markdown file: the Preamble, the tables for Audio and Image, the long-form document from step 3.

Does that last part mean the the long-form document is contained within the new markdown file? Or linked?

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+- As a contributor, I need to know what is the expected data type for the specific property. Is it required? ... so that?
+- As a contributor, I need to know how the data property is transformed and validated.
+- As a contributor writing a provider API script, which data properties do I need to collect?
+- As a contributor writing a provider API script, how do I select data for the specific media property Openverse uses?
+- As a contributor writing a provider script, which format should I convert a media property?
+- As a maintainer, I've found a discrepancy between the type of data property the API returns and what is described in the table (e.g., is `null` for a non-nullable property). Is there an issue to fix this data in the catalog database, or do I need to create a new one?
+- As an API user, I need to know what exactly this media property is and what data points from the providers are used for it. E.g., does the "creator" refer to the person who created the object or took the picture?
+- As a provider of openly-licensed media, I want to know what media properties Openverse collects so that I can change my API responses.
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :40]: I appreciate these user stories a lot, some of them I had not thought of! 💯

**Review by stacimc — COMMENTED** [review]
I'm so excited for this! The sample documentation is such a nice preview of things to come :)

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 1 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :59]: It is very difficult to parse out validation from the `MediaStore` using ast, so I thought that the validation for each property would be manually described in the markdown file.

Do you think that it would be reasonable and useful to track changes in the `MediaStore` (and the `ImageStore` and `AudioStore`) and warn a contributor to update the documentation if necessary? This could be a precommit warning (I don't like this idea because it's very easy to miss; or even if it is possible to add a non-error warning to precommit) or a CI step with a comment on the PR. We cannot use an error instead of a warning for that because the changes to media store might not actually need to have corresponding changes to the documentation.
For implementing this warning, we can use `ast` to get the line numbers for `add_item` and `clean_metadata`, and then check `git diff` to see if any of these lines are affected. I'm not sure what to do about the functions that `clean_metadata` calls, though.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 6 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :40]: Thank you! I've thought of adding the user stories after a couple of weeks thinking about this project, and they actually helped me clarify my thoughts :)

File: documentation/projects/proposals/document_all_media_properties/index.md [hunk 7]
```diff
@@ -1,5 +1,8 @@
 # Document all media properties
 
-```{include} /_toc_templates/flat.md
+```{toctree}
+:titlesonly:
+:glob:
 
+*
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply documentation/projects/proposals/document_all_media_properties/index.md:7]: I think the implementation plan document is not correctly formatted:

```
/home/sara/projects/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:4: WARNING: Document headings start at H2, not H1 [myst.header]
/home/sara/projects/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:8: WARNING: Document headings start at H2, not H1 [myst.header]
/home/sara/projects/openverse/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:13: WARNING: Document headings start at H2, not H1 [myst.header]

```

Notice that the error is in the project proposal. The `#` at the start of that document is missing: 

https://github.com/WordPress/openverse/blob/965b9fab5762078777dd5f58621e0d7a3d99e519/documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md#L1

If the title is fixed there, that resolves one problem.

There is however another very sneaky issue! The implementation plan markdown file's full name is `"20230426-implementation_plan_documenting_media_properties.md "`, with a space at the end :open_mouth:. That appears to be causing some other issues like GitHub (and even my local editor) not treating the file as Markdown.

Another small issue is that the titles for the documents in this directory do not include a date stamp at the front like other documents. E.g.:

https://github.com/WordPress/openverse/blob/697f62f01a32cb7fcf2f4a7627650a113cba40da/documentation/projects/proposals/frontend_ui_state_cookie/20220218-project_proposal.md#L1

The last thing I noticed was that the sample media properties document's title gets rendered directly into the toctree as "Media Properties" without much context that it's meant as a sample document. Perhaps moving that into its own directory with its own index file so that it has an additional heading of "Sample documentation" or something, would be helpful.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 2 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :61]: I think I learned it from @AetherUnbound in her DAG documentation generator 😆 

Step 3 refers to the manually created file (I will refer to it as `media_properties.md` in the revised version of this proposal). It would be saved in the `media_properties_gen` folder inside `/catalog/utilities`, in the same way as `dag_doc_gen`. It would have the data that we cannot devise from code or docstrings.
For ease of parsing, we can have a `media_properties.md` for all the properties, and `general.md` for the Preamble and "postamble". The latter would contain notes for inconsistencies that occur on more than one field (like the encoding problem) or anything else that might be useful but can't be put under another heading.

This is a sample of what I think `media_properties` would look like. 

---

# Title
The title of the work. If blank, uses "This work" for the attribution sentence (This work by creator is licensed with CC BY).

## Shape of the data and Selection criteria
We select the default title returned by the provider. It can be blank. Blank values (whether `None` or empty string `""`) are saved as empty string in the database (TODO: check if this is true).

## Existing data problems
- Some media items had incorrectly encoded titles [^1 - Link to a description of Unicode encoding problem in the "postamble"]. This is compensated for in the Frontend (link to the code that fixes title encoding). This problem has been fixed for the items that have been reingested after some time in 2020, but might still persist for items that were not updated since then. Link to issues for fixing the encoding in the catalog/api/frontend.
- Some Wikimedia titles have a shape of "FILE:xxx.svg". The provider script removes them now, but this is still a problem for items that were ingested earlier. The "FILE" and ".extension" are removed in the frontend (link to the code). Link to the issue to fix it in the API.

---

I would like to link all the issues for fixing the data to the documentation, but I'm not sure if adding them under "Existing data problems" is the best solution for that. There would need to be some mechanism of ensuring that the data is updated after the issue is fixed.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 5 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :64]: Step 4 is referring to the actual documentation page that would be posted to docs.openverse.org This will be auto-generated, and will use the data from steps 1-3 to create something like this:

# Openverse Media Properties

preamble (copied from `general.md`)

## Overview of the media properties

table with the datatype details, possibly with links to the headings below. Autogenerated in steps 1-2.

## In-depth description

### identifier
Description of the `identifier` field. Step 3 reads the manually-written `media_properties.md` file, splits it by the media property, and then writes it here in the correct order.

### title 
description

# Other considerations (postamble)
## Encoding problems
Description of what happened with encoding of tags and titles

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 3 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :74]: This project will add `media_doc_gen` folder to `/catalog/utilities`. This will have the following files:
- `general.md`. Manually written pre- and post- ambles.
- `media_properties.md`. Manually written markdown file that will describe each media property. See examples in the [comment below](https://github.com/WordPress/openverse/pull/1904#discussion_r1179913440)
- `media_doc_generation.py`

It will also add `generate-media-docs` just script that would run `media_doc_generation.py` and create a `media_properties_doc.md` (the name needs to be improved) from the data that is collected in steps 1-3. This will be compared to `/documentation/catalog/media_properties_doc.md` that contains the most up-to-date information, and fail CI if there are differences.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate a markdown page with a table and long-form notes. This script can `diff` with the existing documentation page (created in step 4 below) for the `precommit` check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
```
[the remark this answers is not in the document]
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :53]: I've added a note about the #1894 changes, but forgot to push it :) It would actually make generating the docs much simpler - we would need to only parse one sql file per media type.

File: documentation/projects/proposals/document_all_media_properties/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/index.md:7]: Thank you for the thorough investigation, @sarayourfriend ! I will try to fix the rendering

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 1 above)
Comment (stacimc) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :59]: Totally, I don't see how those sorts of changes could possibly be parsed.

> Do you think that it would be reasonable and useful to track changes in the MediaStore (and the ImageStore and AudioStore) and warn a contributor to update the documentation if necessary?

Very cool idea! A comment on the PR seems ideal and mirrors migration warnings.

> For implementing this warning, we can use ast to get the line numbers for add_item and clean_metadata, and then check git diff to see if any of these lines are affected. I'm not sure what to do about the functions that clean_metadata calls, though.

It might even make sense to just issue the warning if any of the `MediaStore` files are changed at all, as long as it's a warning and not a blocker. What do you think?

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 3 above)
Comment (stacimc) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :74]: Oh I see, this comment really helped me understand the structure here. So there is one, final `media_properties_doc.md` file that’s generated by the `media_doc_generation.py` script, and its structure is:
* manually written preamble (pulled from `general.md`)
* media properties (the script will pull information for each property from docstrings and other files, and then ALSO parse out the manually written information from `media_properties.md` and add them in)
* manually written postamble (pulled from `general.md`)

So the final output is a single markdown file that includes both the manually written info, and the parsed information. Very cool!

Could you add this description of the files to the doc, maybe right before the Outlined Steps? This is really useful!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 5 above)
Comment (stacimc) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :64]: Got it, this comment and [this one](https://github.com/WordPress/openverse/pull/1904/files#r1179918049) totally cleared up my confusion. Thanks!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] AetherUnbound
+- [ ] Krysal ? Dhruv ?
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md :12]:
```suggestion
- [ ] @stacimc 
```

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 1 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md ]: I love the idea of a warning when `MediaStore` files are changed, it greatly simplifies the implementation :) And is still very useful.

File: documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md [hunk 10]
```diff
@@ -1,4 +1,5 @@
-Project proposal for
+# 2023-03-21 Project proposal for
+
 [Document all media properties](https://github.com/WordPress/openverse/issues/412)
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230307-project_proposal.md:3]: This makes the title fall on a second line: 
image

Perhaps we could add this link at the top of the description?

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
+# Implementation Plan: Catalog
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:1]: Nit for consistency (per #1973):
```suggestion
# 2023-04-26 Implementation Plan: Document all media properties in the catalog
```

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+over time. This plan will outline the creation of the documentation page about
+the media properties in the catalog. This will help answer some of the questions
+like:
+
+### User stories
+
+- As a maintainer/contributor, I need to know the list of available properties
+  for each media type in the catalog database. ... so that?
+- As a contributor, I need to know what is the expected data type for the
+  specific property. Is it required? ... so that?
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:40]: I might be unfamiliar with user stories in the traditional sense, is the `... so that?` intended here or was there other information/examples that should be provided?

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 6 above)
Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md ]: Agreed, I love this!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+The information on data types, transformations, and validations will be
+presented as a table and will be extracted from the code. Other information
+about the data selection properties and data discrepancies will be written in a
+separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from
+the DAGs and creates a page with the documentation. We can use the same approach
+to create a documentation page for the media properties. The DAG doc generator
+runs the code inside docker to extract DAG information. In this project, we can
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:67]: Small potential wording change here for clarity:
```suggestion
to create a documentation page for the media properties. The DAG doc generator
runs the code inside docker to extract DAG information because it requires the project dependencies (i.e. `airflow`) in order to parse the DAG. In this project, we can
```

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 8 above)
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md ]: +1 to linking the DDL files in question here!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md  (hunk 2 above)
Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md ]: The idea of separating the pre/postamble into different files is great! That way we don't have to stringify it and reproduce it over and over again with each generation of the file. Do you mind adding all of this description to the IP? It's excellent! 😄

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+for Audio and Image, the long-form document from step 3.
+
+### Information to include in `media_properties.md` for each property
+
+Media types: Image, Audio Short description Names used in the provider scripts,
+if different from the database name Shape of the object (if it is not a simple
+type) Selection criteria Normalization and validation performed in the
+`MediaStore` class and the relevant `Column` Inconsistencies in the database
+data See sample for `identifier` and `tags` in `sample_media_properties.md`
+
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:121]: Was this intended to be a list? It's a little hard to follow in the rendered markdown with it altogether like this 😅

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+presented as a table and will be extracted from the code. Other information
+about the data selection properties and data discrepancies will be written in a
+separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from
+the DAGs and creates a page with the documentation. We can use the same approach
+to create a documentation page for the media properties. The DAG doc generator
+runs the code inside docker to extract DAG information. In this project, we can
+simplify and use the `ast` module to parse static code and extract the
+docstrings to make the checks faster. This will, however, mean that some of the
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:69]: If speed were _really_ a bottleneck, we could always [use rust 😉 🦀](https://docs.rs/python-parser/latest/python_parser/)

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+## Alternative (or additional) Approaches
+
+We could also create a list of classes/functions that are used to transform the
+data and run a `git diff` on them to see if there are any changes. This would be
+a more manual approach and would require more maintenance.
+
+Another option for documenting the media properties is to pass a dataclass
+instead of the named parameters to the `add_item` method of the `ImageStore` and
+`AudioStore`. This would make the documentation closer to code for easier
+updating, adn have another benefit of allowing us to check for required
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:135]:
```suggestion
updating, and have another benefit of allowing us to check for required
```

File: documentation/projects/proposals/document_all_media_properties/sample_documentation/sample_media_properties.md
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/sample_documentation/sample_media_properties.md:1]: This is excellent!! ✨

**Review by AetherUnbound — COMMENTED** [review]
Thanks for writing this implementation plan Olga, this is fantastic stuff! I've got a few questions/suggestions in addition and in comport with Staci.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md (hunk 15 above)
Comment (obulat) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:69]: Wow! That's interesting, but ... I don't think it _really_ is a bottleneck :)

**Comment by obulat** [comment]
I'm moving the IP into the Revision Round and will be drafting the PR while I make updates based on feedback.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md (hunk 12 above)
Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:40]: Got it, thanks!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md (hunk 15 above)
Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md]: I don't think it is either ha, I just got curious!

**Comment by obulat** [comment]
I need advice on where to put the resulting file, and how to call it. Below are some options that I've thought of:
Options for URLs:

- https://docs.openverse.org/media/media_properties
- https://docs.openverse.org/media/media_in_openverse
- https://docs.openverse.org/data/media_properties

Options for titles:

- "Openverse Media Properties"
- "Media in Openverse"
- "Media Properties"

**Comment by obulat** [comment]
I've updated the proposal and moved it to the decision round, @AetherUnbound , @stacimc 

The changes added are:
- Clarified what inputs will be used to generate the final document, and added more details to the samples
- Rewrote the steps to actually refer to the workstreams that can be done in parallel, instead of describing the steps of the input file parsing.

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 17]
```diff
@@ -61,164 +69,194 @@ presented as a table and will be extracted from the code. Other information
...
-the shape and kind of data we expect to have for the property, how to select it
-from the provider, and any inconsistencies between the data we have in the
-database and the data we expect to have. This file is parsed in step 3 and the
-information is added to the final page on the documentation site.
+The result of this project will be a markdown page with a table of media
+properties and a long-form description of each property.
 
-### `media_properties_documentation.md`
+The page will be published on
+https://docs.openverse.org/openverse_media_properties_documentation.html with
```
Comment (AetherUnbound) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:78]: Based on L82, wouldn't this be?
```suggestion
https://docs.openverse.org/catalog/media_properties_documentation.html with
```

And I think we were discussing putting this under the `/meta`[^1] hierarchy rather than `/catalog`, since we'll have definitions for the other services as well in future IPs that might be useful to have colocated in in the documentation. What do you think?

[^1]: https://github.com/WordPress/openverse/pull/1904#issuecomment-1536588260

**Review by AetherUnbound — APPROVED** [review]
I am so excited for this project! 💃🏼 ✨ Your updates look great, the only notes I have are on a potentially duplicated user story and the final URL/location for the media properties file. Great work!!

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @AetherUnbound
+- [ ] @stacimc
```
Comment (stacimc) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:11]:
```suggestion
- [x] @stacimc
```

**Review by stacimc — APPROVED** [review]
So sorry for the delay @obulat -- your updates look great and address all my questions :)

File: documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md [hunk 19]
```diff
@@ -0,0 +1,258 @@
+# 2023-04-26 Implementation Plan: Document all media properties in the catalog
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @AetherUnbound
```
Comment (obulat) [inline documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md:10]:
```suggestion
- [x] @AetherUnbound
```

Comment (AetherUnbound) [reply documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md]: Oops, sorry for not doing this on my review!
